packages/react/src/combobox/root/ComboboxRoot.tsx

```diff
@@ -0,0 +1,1135 @@
+  useClick,
+} from '../../floating-ui-react';
+import { contains, getTarget } from '../../floating-ui-react/utils';
+import {
+  BaseOpenChangeReason,
+  translateOpenChangeReason,
+} from '../../utils/translateOpenChangeReason';
+import {
+  ComboboxFloatingContext,
+  ComboboxDerivedItemsContext,
+  ComboboxRootContext,
+  ValueChangeReason,
+} from './ComboboxRootContext';
+import { selectors, type State as StoreState } from '../store';
+import { useOpenChangeComplete } from '../../utils/useOpenChangeComplete';
+import { CompositeList } from '../../composite/list/CompositeList';
+import { useFieldRootContext } from '../../field/root/FieldRootContext';
+import { useFieldControlValidation } from '../../field/control/useFieldControlValidation';
+import { useFormContext } from '../../form/FormContext';
+import { useField } from '../../field/useField';
+import { useBaseUiId } from '../../utils/useBaseUiId';
+import {
+  type ComboboxGroup,
+  isGroupedItems,
+  defaultGroupFilter,
+  stringifyItem,
+  createCollatorItemFilter,
+  createSingleSelectionCollatorFilter,
+} from './utils';
+import { useFilter } from './utils/useFilter';
+import { serializeValue } from '../../utils/serializeValue';
+import { useTransitionStatus } from '../../utils/useTransitionStatus';
+import { EMPTY_ARRAY } from '../../utils/constants';
+import { useOpenInteractionType } from '../../utils/useOpenInteractionType';
+import { HTMLProps } from '../../utils/types';
+
+export function ComboboxRoot<Item = any, Mode extends SelectionMode = 'none'>(
+  props: ComboboxRootConditionalProps<Item, Mode>,
+): React.JSX.Element {
+  const {
+    id: idProp,
+    onOpenChangeComplete: onOpenChangeCompleteProp,
+    defaultSelectedValue = null,
+    selectedValue: selectedValueProp,
+    onSelectedValueChange,
+    defaultInputValue = '',
+    inputValue: inputValueProp,
+    selectionMode = 'none',
+    onItemHighlighted: onItemHighlightedProp,
+    name: nameProp,
+    disabled: disabledProp = false,
+    readOnly = false,
+    required = false,
+    inputRef: inputRefProp,
+    cols = 1,
+    items,
+    filter: filterProp,
+    openOnInputClick = true,
+    itemToString,
+    itemToValue,
+    virtualized = false,
+    fillInputOnItemPress = true,
+    modal = false,
+    clearInputOnCloseComplete = false,
+  } = props;
+
+  const { clearErrors } = useFormContext();
+  const {
+    setDirty,
+    validityData,
+    validationMode,
+    setControlId,
+    setFilled,
+    name: fieldName,
+    disabled: fieldDisabled,
+  } = useFieldRootContext();
+  const fieldControlValidation = useFieldControlValidation();
+
+  const id = useBaseUiId(idProp);
+  const disabled = fieldDisabled || disabledProp;
+  const name = fieldName ?? nameProp;
+  const multiple = selectionMode === 'multiple';
+
+  const frame = useAnimationFrame();
+
+  useIsoLayoutEffect(() => {
+    setControlId(id);
+    return () => {
+      setControlId(undefined);
+    };
+  }, [id, setControlId]);
```

Comment:
This pattern is repeated often, would it make sense to add it to `useBaseUiId(idProp, setControlId)`?